coredocument/read_acls.go

```diff
@@ -110,7 +144,100 @@ func isPeerInRole(role *coredocumentpb.Role, peer identity.CentID) bool {
 	return false
 }
 
-// peerValidator return the
+// peerValidator returns the ReadAccessValidator tp verify peer.
 func peerValidator() ReadAccessValidator {
 	return readAccessValidator{}
 }
+
+// nftValidator returns the ReadAccessValidator for nft owner verification.
+func nftValidator(tr TokenRegistry) ReadAccessValidator {
+	return readAccessValidator{tokenRegistry: tr}
+}
+
+// NFTOwnerCanRead checks if the nft owner/peer can read the document
+// Note: signature should be calculated from the hash which is calculated as
+// keccak256("\x19Ethereum Signed Message:\n"${message length}${message}).
+func (r readAccessValidator) NFTOwnerCanRead(
+	cd *coredocumentpb.CoreDocument,
+	registry common.Address,
+	tokenID []byte,
+	signature string,
+	peer identity.CentID) error {
+
+	// check if the peer can read the doc
+	if r.PeerCanRead(cd, peer) {
+		return nil
+	}
+
+	// check if the nft is present in read rules
+	found := findRole(cd, coredocumentpb.Action_ACTION_READ, func(role *coredocumentpb.Role) bool {
+		return isNFTInRole(role, registry, tokenID)
+	})
+
+	if !found {
+		return errors.New("nft missing")
+	}
+
+	// get the owner of the NFT
+	owner, err := r.tokenRegistry.OwnerOf(registry, tokenID)
+	if err != nil {
+		return errors.New("failed to get NFT owner: %v", err)
+	}
+
```

Comment:
By default we should check if token owner == sender and approve if true. This is the minimum requirement. I would say more important than the signature below.

Reply:
Thinking more about this actually:

If the NFT owner is not a centrifuge identity, then what is the likelihood that they could easily create the signature. Let's assume you do run a centrifuge node and want to read a document, you would also have to make sure that the ethereum private key that owns the document is also configured on the node as a private key so the node that makes the request can generate the signature. I don't know how likely that is or what the benefit would be in this over just having an identity own the NFT. If the node does not have the private key of the account, then you would have to manually construct the signature and provide the signature as an additional argument to the rest api when making the read request. The usability of this to do this manually is definitely less than ideal and even in a fully integrated situation it's still questionable how often this is practical.

I do think this might be something we want to support in the future, but I really don't see this as a very convenient way for users to read documents in the beginning. Just checking that it is a centrifuge identity should definitely be supported.